### PR TITLE
torch.linalg.lu: improve heuristics for the cuSOLVER vs cuBLAS switch

### PR DESCRIPTION
**Pull request by nikitaved**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #185344

The data gathered across 4 GPUs (A100, H100, L40, RTX5090), with about 3800 points in total. 
Fixes https://github.com/pytorch/pytorch/issues/181999. Also, helps us with better differentiating when MAGMA outperforms cuBLAS/cuSOLVER -- useful information for the ongoing deprecation process.

batch=2 is the noisy area -- L40/RTX5090 and H100 tend do disagree there, and one needs to make GPU-aware decisions in this region. However, the goal is to be mostly GPU-agnostic.   

This is the analysis of the heuristic:
```
======================================================================                                             
LU Backend Heuristic Validation                                                                                    
======================================================================                                                  
                                                                                                                        
Overall accuracy: 3696/3792 (97.5%)                                                                                     
                                                                                                                        
--- Accuracy by GPU ---                                                                                                     
      a100: 944/955 (98.8%)                                                                                                 
      h100: 925/955 (96.9%)                                                                                                 
       l40: 914/947 (96.5%)                                                                                                     
   rtx5090: 913/935 (97.6%)                                                                                                     
                                                                                                                                
--- Accuracy by dtype ---                                                                                                           
       float32: 940/956 (98.3%)                                                                                                     
       float64: 915/949 (96.4%)                                                                                                     
     complex64: 936/949 (98.6%)                                                                                                         
    complex128: 905/938 (96.5%)                                                                                                         
                                                                                                                                        
--- Accuracy by batch size ---                                                                                                              
  batch=  2: 613/640 (95.8%)                                                                                                                
  batch=  4: 604/640 (94.4%)                                                                                                                
  batch=  8: 624/639 (97.7%)                                                                                                                   
  batch= 16: 628/635 (98.9%)                                                                                                                   
  batch= 32: 618/626 (98.7%)                                                                                                                   
  batch= 64: 609/612 (99.5%)                                                                                                                      
                                                                                                                                                  
--- Wrong predictions: 96 cases ---                                                                                                                  
     GPU        dtype batch      N  predicted     actual slowdown                                                                                    
-----------------------------------------------------------------                                                                                      
    h100   complex128    64   1536     cublas   cusolver    1.81x                                                                                        
    a100   complex128    64   1536     cublas   cusolver    1.70x                                                                                        
     l40      float64     2     65     cublas   cusolver    1.59x                                                                                         
    h100   complex128     2     96   cusolver     cublas    1.50x                                                                                           
 rtx5090   complex128     4    193     cublas   cusolver    1.47x                                                                                           
     l40   complex128     4    193     cublas   cusolver    1.46x                                                                                            
     l40      float64     2     63     cublas   cusolver    1.43x                                                                                                 
     l40      float64     2     64     cublas   cusolver    1.41x                                                                                                 
     l40      float64     2     48     cublas   cusolver    1.32x                                                                                                 
    h100   complex128     8    352   cusolver     cublas    1.29x                                                                                                 
    h100      float64     2     96   cusolver     cublas    1.26x                                                                                                 
     l40      float64     4    256     cublas   cusolver    1.26x                                                                                                  
     l40      float64     4    352     cublas   cusolver    1.26x                                                                                                     
 rtx5090   complex128     4    192     cublas   cusolver    1.25x                                                                                                     
 rtx5090   complex128     4    191     cublas   cusolver    1.25x                                                                                                     
     l40      float64     4    224     cublas   cusolver    1.25x                                                                                                     
     l40      float64     4    193     cublas   cusolver    1.24x                                                                                                     
    h100      float32     4    288   cusolver     cublas    1.24x                                                                                                        
     l40   complex128     4    192     cublas   cusolver    1.24x                                                                                                        
     l40   complex128     4    191     cublas   cusolver    1.24x                                                                                                        
     l40      float64     4    255     cublas   cusolver    1.22x                                                                                                             
 rtx5090      float64     4    352     cublas   cusolver    1.22x                                                                                                                
     l40      float32     4    255     cublas   cusolver    1.22x                                                                                                             
 rtx5090      float64    16   1025     cublas   cusolver    1.21x                                                                                                             
 rtx5090      float64    16   1024     cublas   cusolver    1.18x                                                                                                             
 rtx5090      float64    16   1023     cublas   cusolver    1.18x                                                                                                             
     l40      float64    16   1025     cublas   cusolver    1.17x                                                                                                             
     l40      float64    16   1024     cublas   cusolver    1.16x                                                                                                                
    h100      float64     2    127   cusolver     cublas    1.16x                                                                                                                
     l40      float64    16   1023     cublas   cusolver    1.16x                                                                                                                
    h100   complex128     4    224   cusolver     cublas    1.15x                                                                                                                
    a100   complex128     2     96   cusolver     cublas    1.15x                                                                                                                       
    h100      float32     2    160   cusolver     cublas    1.15x                                                                                                                       
    a100   complex128     8    352   cusolver     cublas    1.15x                                                                                                                              
    h100      float64     4    383   cusolver     cublas    1.15x                                                                                                                       
     l40      float32     4    256     cublas   cusolver    1.14x                                                                                                                       
    h100   complex128     2    128   cusolver     cublas    1.14x                                                                                                                       
    h100      float64     4    384   cusolver     cublas    1.13x                                                                                                                       
    h100    complex64     4    288   cusolver     cublas    1.12x                                                                                                                       
    h100      float64     2    128   cusolver     cublas    1.12x                                                                                                                              
    h100    complex64     2    160   cusolver     cublas    1.12x                                                                                                                              
    h100      float32     8    480   cusolver     cublas    1.11x                                                                                                                                     
     l40   complex128     2     65     cublas   cusolver    1.11x                                                                                                                              
    h100   complex128     2    127   cusolver     cublas    1.11x                                                                                                                              
 rtx5090      float64     2     65     cublas   cusolver    1.11x                                                                                                                              
    h100   complex128     8    384   cusolver     cublas    1.10x                                                                                                                              
     l40   complex128     2     63     cublas   cusolver    1.10x                                                                                                                              
 rtx5090      float64     4    193     cublas   cusolver    1.10x                                                                                                                                     
 rtx5090    complex64     4    256     cublas   cusolver    1.10x                                                                                                                                     
     l40    complex64     4    256     cublas   cusolver    1.09x                                                                                                                                     
     l40    complex64     4    255     cublas   cusolver    1.09x                                                                                                                                        
 rtx5090      float64     4    256     cublas   cusolver    1.09x                                                                                                                                     
 rtx5090    complex64     4    255     cublas   cusolver    1.09x                                                                                                                                     
 rtx5090      float64     4    224     cublas   cusolver    1.08x                                                                                                                                     
    h100   complex128     8    383   cusolver     cublas    1.08x                                                                                                                                     
    a100   complex128    16    511     cublas   cusolver    1.08x                                                                                                                                     
     l40   complex128     2     64     cublas   cusolver    1.08x                                                                                                                                        
    a100      float32     4    288   cusolver     cublas    1.07x                                                                                                                                        
    h100      float64     2    129   cusolver     cublas    1.07x                                                                                                                                        
    a100   complex128    32    767     cublas   cusolver    1.07x                                                                                                                                        
    h100      float32     8    513   cusolver     cublas    1.07x                                                                                                                                               
 rtx5090      float64     4    255     cublas   cusolver    1.06x                                                                                                                                               
 rtx5090    complex64     8    385     cublas   cusolver    1.06x                                                                                                                                               
     l40      float32     2    127     cublas   cusolver    1.06x                                                                                                                                                      
    a100   complex128    32    769     cublas   cusolver    1.05x                                                                                                                                               
     l40      float32     2    129     cublas   cusolver    1.05x                                                                                                                                               
     l40   complex128     8    352   cusolver     cublas    1.05x                                                                                                                                               
    h100      float32     8    512   cusolver     cublas    1.05x                                                                                                                                               
     l40    complex64     2    129     cublas   cusolver    1.04x                                                                                                                                               
 rtx5090   complex128    64   1536     cublas   cusolver    1.04x                                                                                                                                                      
 rtx5090      float32     2    160   cusolver     cublas    1.04x                                                                                                                                                      
 rtx5090    complex64     4    257     cublas   cusolver    1.04x                                                                                                                                                              
    a100      float64     4    352     cublas   cusolver    1.03x                                                                                                                                                      
    h100      float64     8    767   cusolver     cublas    1.03x                                                                                                                                                      
     l40      float32     4    257     cublas   cusolver    1.03x                                                                                                                                                      
     l40      float64     4    257     cublas   cusolver    1.03x                                                                                                                                                      
    h100      float64     2    160   cusolver     cublas    1.03x                                                                                                                                                      
    h100      float64     8    768   cusolver     cublas    1.02x                                                                                                                                                              
     l40   complex128     4    127     cublas   cusolver    1.02x                                                                                                                                                              
    h100    complex64     8    480   cusolver     cublas    1.02x                                                                                                                                                              
    a100    complex64    32    767     cublas   cusolver    1.02x                                                                                                                                                              
 rtx5090      float64    32   1536     cublas   cusolver    1.02x                                                                                                                                                                   
    h100   complex128    32    769     cublas   cusolver    1.02x                                                                                                                                                              
     l40   complex128     4    128     cublas   cusolver    1.02x                                                                                                                                                              
 rtx5090   complex128     8    255     cublas   cusolver    1.02x                                                                                                                                                              
    h100   complex128     2    129   cusolver     cublas    1.01x                                                                                                                                                              
    a100    complex64    32    769     cublas   cusolver    1.01x                                                                                                                                                              
    h100   complex128    32    767     cublas   cusolver    1.01x                                                                                                                                                                      
    a100      float64     2     96   cusolver     cublas    1.01x                                                                                                                                                                   
 rtx5090      float32     4    255     cublas   cusolver    1.01x                                                                                                                                                                   
     l40   complex128    32   1025   cusolver     cublas    1.01x                                                                                                                                                                   
    h100      float32     8    511   cusolver     cublas    1.01x                                                                                                                                                                   
     l40      float32     4    224     cublas   cusolver    1.01x                                                                                                                                                                   
    h100      float32     2    192   cusolver     cublas    1.00x                                                                                                                                                                      
 rtx5090   complex128     8    256     cublas   cusolver    1.00x                                                                                                                                                                      
     l40    complex64     2    127     cublas   cusolver    1.00x                                                                                                                                                                      
                                                                                                                                                                                                                                       
Severity breakdown of wrong predictions:                                                                                                                                                                                                  
  < 1.5x slowdown (mild):     93                                                                                                                                                                                                          
  1.5-3x slowdown (moderate):  3                                                                                                                                                                                                          
  > 3x slowdown (severe):      0                         
```

Authored with Claude.

cc @ptrblck @msaroufim @eqy @tinglvv @nWEIdia @csarofeen @jianyuh @mruberry @walterddr @xwang233 @Lezcano

### CODE REVIEW

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 1]
```diff
@@ -858,6 +858,100 @@ static void lu_factor_batched_magma(const Tensor& input, const Tensor& pivots, c
...
+  // advantage over cuSOLVER persists to larger N at high batch. Empirically
+  // N^2/batch at the crossover roughly doubles each time batch doubles.
+
+  // complex128 also uses batch^1.5 scaling empirically, though its crossover
+  // ratios are less regular. The lower multiplier (5200 vs 16600) reflects
+  // that cuSOLVER overtakes cuBLAS at smaller N for complex128.
+  //
+  // FIXME: test on Blackwell with float32 via half emulation enabled (CUDA 13.2+).
+  //
+  // Authored with Claude.
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: The benchmark summary and the rule are useful future-reader context, but `Authored with Claude` does not help explain or maintain the heuristic. It's your code still.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 2]
```diff
@@ -858,6 +858,100 @@ static void lu_factor_batched_magma(const Tensor& input, const Tensor& pivots, c
...
 
+#ifdef USE_LINALG_SOLVER
+enum class SolverBackend : char {
+  CUSOLVER,
+  CUBLAS
+};
+#ifndef USE_ROCM
+namespace {
+
+  // Based on benchmarks across H100, A100, L40, RTX5090 with about 3200 points:
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: This says the rule is based on about 3200 points, while the PR description says about 3800 points and the validation table totals 3792.

Can we make these numbers agree, or phrase this as the exact validation set used for the constants? This heuristic is benchmark-derived, so the provenance needs to be easy to trust later.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp [hunk 3]
```diff
@@ -858,6 +858,100 @@ static void lu_factor_batched_magma(const Tensor& input, const Tensor& pivots, c
...
+  // float64 needs super-linear scaling (batch^1.5) because cuBLAS's cost
+  // grows more slowly with N for float64 than for other dtypes - its
+  // advantage over cuSOLVER persists to larger N at high batch. Empirically
+  // N^2/batch at the crossover roughly doubles each time batch doubles.
+
+  // complex128 also uses batch^1.5 scaling empirically, though its crossover
+  // ratios are less regular. The lower multiplier (5200 vs 16600) reflects
+  // that cuSOLVER overtakes cuBLAS at smaller N for complex128.
+  //
+  // FIXME: test on Blackwell with float32 via half emulation enabled (CUDA 13.2+).
```
Comment (IvanYashchuk) [inline aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: What Blackwell? The PR description says 5090 was used, it's also Blackwell.

**Review by IvanYashchuk — COMMENTED** [review]
The direction is right. The old `m >= 512` cutoff is too coarse for batched square LU, and the benchmark table in the PR is a much better basis for the cuBLAS/cuSOLVER choice.

Are you planning to add user-controlled overwrite of this choice? It could probably be done through https://docs.pytorch.org/docs/2.12/backends.html#torch.backends.cuda.preferred_linalg_library adding a cublas option there.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp (hunk 2 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: New data came in, need to update.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp (hunk 3 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: It is optimized for inference, hence fp32/fp64 perf is not great, unless they are emulated by half precision, but we need to enable it plus CUDA 13.2.

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp (hunk 1 above)
Comment (nikitaved) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: Will remove it. All the dirty work is done by me :)

**Comment by nikitaved** [comment]
Thank you, @IvanYashchuk , for your comments. I am not sure that having cublas as a separate backend is justifiable just yet -- LU, QR, LU_solve, LSTSQ -- the benchmarks are old there, they need a re-eval (and heuristics appear to be quite nontrivial for some of them).

File: aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp (hunk 3 above)
Comment (johannesz-codes) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: It's available from 13.0. Also cuBLAS enables it by default if I understand the docs right.

Comment (nikitaved) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: Thank you, @johannesz-codes , I will update the comment. Indeed, fp64 is there since 13.

Comment (nikitaved) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: I will double-check the heuristic for such cases then. This is a TODO.

Comment (nikitaved) [reply aten/src/ATen/native/cuda/linalg/BatchLinearAlgebra.cpp]: OK, no need to modify things when testing cusolver/cublas on CUDA 13.2 each with/without fp64 emulation enabled. The fp64 advantage really kicks in at quite large sizes, very much away from the captured decision boundary.

**Comment by nikitaved** [comment]
@IvanYashchuk , could you give it another review pass?

**Review by IvanYashchuk — APPROVED** [review]
